src/Functions/parseDateTime.cpp

```diff
@@ -684,7 +746,16 @@ namespace
                 if (result.has_value())
                 {
                     if (result = datetime.buildDateTime(time_zone); result.has_value())
-                        res_data[i] = static_cast<UInt32>(*result);
+                    {
+                        if constexpr (parseDateTime64)
+                        {
+                            const DataTypeDateTime64 * datatime64_type = checkAndGetDataType<DataTypeDateTime64>(removeNullable(result_type).get());
```

Comment:
The stuff computed in l. 752-753 doesn't change between rows. We should pull it out of the loop.